### DISCUSSION TITLE
get the most outer element's id?

### DISCUSSION THREAD

**Discussion by retsyo** [zauberzeug/nicegui#2144] [Q&A] [answered]
### Question

In the [javascript code which exports webpage to pdf](https://blog.bitsrc.io/how-to-export-webpage-to-pdf-using-javascript-html2pdf-and-jspdf-6cdd549618c), we must specify which element should be exported, for example

```javascript
var element = document.getElementById('divToExport');
```

for a NiceGUI application, how to get the `id` of the current, whole page? or get the most outer element object in python and pass its to JavaScript's `document.getElement()`?
 
```python
# what is the id of this whole application?

from nicegui import ui

ui.button()
ui.label()
ui.run()
```

After observe some different NiceGUI applications, I conclude that it is safe to use `'c0'` as the `ID` of the complete page now, for example
```html
<div class="q-layout q-layout--standard nicegui-layout" tabindex="-1" id="c0" style="min-height: 1325px;">
```
But I do not know whether this value is valid in different NiceGUI versions.  

Thanks.

**Comment by falkoschindler** [comment, marked as answer]
Yes, right now "c0" is the ID of the root element, the QLayout. You can also use the class "nicegui-layout" to address this element.
But even though I don't see why this should change in the future, it _could_. There have been discussions about leaving the `id` attribute to be set by the user, which would require NiceGUI to use a different attribute to identify elements. But this isn't on our roadmap for the foreseeable future.

Just to mention some alternatives to hard-coding the ID, you can get the "number" of the root element like this:
```py
context.get_client().layout.id
```
Note that this is the element ID in Python, which is an integer. By convention the HTML ID has the prefix "c", mainly because HTML doesn't like pure integer IDs.

But in case the recipe for converting Python IDs to HTML IDs changes, you can query the root element like this:
```py
async def get_id():
    print(await ui.run_javascript(f'getElement({context.get_client().layout.id}).$el.id'))
ui.button('Get ID', on_click=get_id)
```
This leverages NiceGUI's `getElement()` function, which accepts an integer ID and returns the Vue component.
